File: container2/leetcodely/python/ds/binary_search_tree.py
class BinarySearchTree:

    def __init__(self):
        self.root = None
        self.size = 0

    def insert(self, item):
        self.root = self._insert(self.root, item)
        self.size += 1

    def _insert(self, root, item):
        if not root:
            root = TreeNode(item)
            return root
        if item < root.val:
            root.left = self._insert(root.left, item)
        else:
            root.right = self._insert(root.right, item)
        return root

    def find(self, item):
        return self._find(self.root, item)

    def _find(self, root, item):
        if not root:
            return False
        if item > root.val:
            return self._find(root.right, item)
        elif item < root.val:
            return self._find(root.left, item)
        else:
            return True

    def delete(self, key):

        def _delete(root, key):
            if not root:
                return
            if key > root.val:
                root.right = _delete(root.right, key)
            elif key < root.val:
                root.left = _delete(root.left, key)
            else:
                if not root.right:
                    return root.left
                else:
                    p = root.right
                    while p.left:
                        p = p.left
                    root.val = p.val
                    root.right = _delete(root.right, p.val)
            return root
        return _delete(self.root, key)


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


if __name__ == '__main__':
    bst = BinarySearchTree()
    bst.insert(12)
    bst.insert(9)
    bst.insert(13)
    print(bst.size)
    print(bst.find(11))

File: container2/leetcodely/javascript/ds/binarySearchTree.js
const Queue = require('./queue');

class BinarySearchTree{
    constructor(){
        this.root = null;
        this.size = 0;

    }

    insert(item){
        this.root = this.insertHelper(this.root, item);
        this.size ++;
    }
    insertHelper(root, item){
        if(root === null){
            root = new Node(item);
            return root;
        }
        if(item < root.val){
            root.left = this.insertHelper(root.left, item);
        }else{
            root.right  = this.insertHelper(root.right, item);
        }
        return root;

    }
    contains(item){
        return this.containsHelper(this.root, item);

    }
    containsHelper(root, item){
        if(root === null){
            return false;
        }
        if (item < root.val){
            return this.containsHelper(root.left, item);
        }else if(item > root.val){
            return this.containsHelper(root.right, item);
        }else{
            return true;
        }

    }
    remove(item){

    }
    balance(){

    }
    preOrder(){
        return this.preOrderhelper(this.root);

    }
    preOrderhelper(root){
        if(root === null){
            return;
        }
        console.log(root.val);
        this.preOrderhelper(root.left);
        this.preOrderhelper(root.right);

    }
    postOrder(){
        this.postOrderHelper(this.root);

    }
    postOrderHelper(root){
        if(root === null){
            return;
        }
        this.postOrderHelper(root.left);
        this.preOrderhelper(root.right);
        console.log(root.val);
    }
    inorder(){
        this.inorderHelper(this.root);

    }
    inorderHelper(root){
        if(root === null){
            return;
        }
        this.inorderHelper(root.left);
        console.log(root.val);
        this.inorderHelper(root.right);

    }
    levelOrderTraversal(){
        const queue = new Queue();
        queue.enqueue(this.root);
        while (!queue.isEmpty()){
            const item = queue.dequeue();
            console.log(item.val);
            if(item.left !== null){
                queue.enqueue(item.left);
            }
            if(item.right !== null){
                queue.enqueue(item.right);
            }
        }
    }
}
class Node{
    constructor(item){
        this.val = item;
        this.left = null;
        this.right = null;
    }
}

module.exports = BinarySearchTree;